### ISSUE TITLE
LevelDB performance optimizations

### ISSUE THREAD

**Issue by dktapps** [pmmp/PocketMine-MP#2925]
We've currently barely scratched the surface of performance improvements that can be made to world I/O with LevelDB.

LevelDB is extremely nice (much nicer than Anvil) because it allows reading/writing parts of chunks on disk without accessing the entire thing. This is hugely beneficial because:
- We currently save the entire chunk to disk when any part of the chunk is marked as dirty. This is incredibly wasteful when we take into consideration that there can be up to 16 subchunks in a chunk, and there's a good chance that only a small number of them (perhaps only one) has changed.
- We save the entire chunk to disk when a chunk contains entities. This is also incredibly wasteful because the terrain is by far the biggest part of the chunk, and with LevelDB it's possible to just save entities alone without the entire chunk.

The bottom line here is that because of the modularity of LevelDB chunks, it's possible to get substantial I/O performance improvements simply by being more selective about the data we read and write.

**Comment by dktapps** [comment]
It's also possible to optimise autosaves by writing all modified data into a cached write batch during autosave and then applying such a batch in one sweep. This would require some changes to the internal architecture of world providers.

**Comment by dktapps** [comment]
Since this was written, chunk saving has been split into 3 parts:

1) Terrain is now only written if modified.
2) Biomes are only written if modified. (This will become significantly more important in 1.18 when 3D biomes arrive.)
3) Entities and tiles are always written.

Each of these steps happens independently.

Implementing terrain writing on a per-subchunk basis might be quite a bit more complicated, though.
